Added PlayerStats packet and VS2015 support

danil179 opened the pull request:
## Resolve #584:
Currently VS2015 has errors while compiling GameServerLib, it is solved by one nuget package.

## Progress #359:
Also PlayerStats (in another branch it called HeroStats) packet was added (not yet implemented or sent at all).

Now scoreboard starting to work, remained to change the stats when needed and also add some missing stats.

### Thanks to loldevs/leaguespec wiki for the specifications of the various fields in this packet.

## Resolve #620:
Now player get gold on kill. (see 73df67de42d7266a6c6e6dafeb5e6042c6e47540)

danil179 committed:
Fixing compiling on VS2015

danil179 committed:
Added PlayerStats packet

danil179 committed:
Tabs to spaces fix

danil179 committed:
Added champion stats

danil179 committed:
Implemented Player/Hero Stats packet

danil179 committed:
Fixed null reference on champion and added sending of packet, tested on one user and it now work

danil179 committed:
Fixed PlayerStats cs not in csproj file

danil179 committed:
Now stats updating as should be (only kills,minions,deaths)

danil179 commented:
Now scoreboard should work (besides items).

danil179 committed:
tested on multiplayer, bug fixed - now kills updated in multiplayer

danil179 commented:
Tested on multiplayer (2 players) and it seems to work, only kills have some bugs.

PR ready for merge.

xTeJk commented:
@danil179 What do you mean by saying "kills have some bugs"?

danil179 commented:
@tejkuu sometimes, kills count wrong with minions, and also there are some bugs if the two killed at the same time, bugs like these.

HarmfulBreeze committed:
Merge branch 'indev' into indev

HarmfulBreeze commented on GameServerLib/Logic/GameObjects/AttackableUnits/AI/Champion.cs:
```diff
@@ -17,6 +17,8 @@ public class Champion : ObjAiBase
         public float ChampionGoldFromMinions { get; set; }
         public RuneCollection RuneList { get; set; }
         public Dictionary<short, Spell> Spells { get; private set; } = new Dictionary<short, Spell>();
+        public ChampionStats ChampStats = new ChampionStats();
```
Make it private

danil179 replied:
But minion class should change some of the stats, because for example when minion die some of the stats of the champion should be changed.

It's possible to make functions for each change (something like 20 functions), but I saw that Stats in AttackableUnit are also Public so I kept ChampStats public as well.

MythicManiac replied:
In cases like that you add a function to the variable holder `Champion in this case` which does what you want, and call that from the minion for example. Generally it's bad practice to modify another object's variables.

danil179 replied:
Ok, I'll write OnKillMinion function in Champion.cs to do that, and make ChampStats property with no set (public to read).

HarmfulBreeze replied:
@MythicManiac I only half-agree with you on that one, C# and properties were especially made for this. With proper synchronization you can do good stuff

HarmfulBreeze commented on GameServerLib/Logic/GameObjects/Stats/ChampionStats.cs:
```diff
@@ -0,0 +1,112 @@
+namespace LeagueSandbox.GameServer.Logic.GameObjects.Stats
+{
+    [StructLayout(LayoutKind.Explicit)]
+    public class ChampionStats
```
Use properties, don't expose public fields

danil179 replied:
From what I checked there is a problem with applying FieldOffsetAtrribute to properties (see here: https://stackoverflow.com/a/28505211) - I can try to check if that's work but I dont think so.

The solution for that will be to write some function that serializes the data, but that's sort of redundancy because all we want is to write the variables in their offset.

MythicManiac replied:
If you convert this to a struct it should be fine to have public non-properties. Is there anything preventing this from being a struct?

danil179 replied:
No, but then I think I should write serializer class for the GetBytes method, or may I keep it in the struct?

MythicManiac replied:
Feel free to keep it in the struct as far as I'm concerned

danil179 replied:
There is a problem after changing the class to struct, everytime you want to change variable in the struct you need to create a new struct (see here: https://stackoverflow.com/questions/1747654/cannot-modify-the-return-value-error-c-sharp), that's beacuse the struct is a value type not like class.

MythicManiac replied:
Can you serialize private variables? If so, the "proper" way to go about this would be
```CS
[FieldOffset(0)]
private int _assists = 0;
public int Assists { get { return _assists; } set { _assists = value; } }
```
however, that would be redundant.

I think it's fine to deviate from our normal style guide in this case unless someone else has a better suggestion. Just make sure to comment in the beginning of the class that it's exempt from the style guide.

HarmfulBreeze replied:
I like this serializing solution. Looks redundant but that's exactly how autoprops work and it'd provide proper encapsulation. So looks good to me

danil179 replied:
So what to implement?
the private variable+public property solution?
or keep it as is?

MythicManiac replied:
Sorry for the confusion 😄, I meant that it's fine to keep it as it is (which deviates from our styleguide), just add a commet mentioning that it's been discussed and it's fine to deviate from the style guide (you can link this PR for example).

HarmfulBreeze reviewed (changes requested):
c.f. my comments
Also the cause of the other issues you mentioned doesn't seem to be related to your PR specifically, so it looks ok to me to merge it even with these bugs

danil179 committed:
Added OnKillChampion

danil179 committed:
Solved problems with the merge

danil179 commented:
Completed the changes, ready for merge.

HarmfulBreeze commented on GameServerLib/Logic/GameObjects/AttackableUnits/AttackableUnit.cs:
```diff
@@ -114,31 +114,7 @@ public virtual void Die(AttackableUnit killer)
-                    cKiller.ChampionGoldFromMinions = 0;
-                    cKiller.KillDeathCounter += 1;
-                }
+                cKiller.OnKillMinion(this as Minion);
```
This can be dangerous as ``this as Minion`` may be null, therefore it may throw an NPE when you kill any ``AttackableUnit`` that is not a minion
I'd say, just do something like 
```cs
if (this is Minion) {
    cKiller.OnKillMinion((Minion) this);
}

HarmfulBreeze replied:
Also this OnKillMinion code should go to ``Minion.Die`` probably

danil179 replied:
but OnKillMinion changes the state of the champion, if we want that hero stats will be private to set that is needed that this function will be in champion class, also I think that is bad to call ```OnKillMinion``` from ```Minion.Die```, and it is better to call it from where we call ```Minion.Die```.

I'll add a check for the type but in champion function, and will change the signature it to ```OnKill (AtttackableUnit killed)```.

HarmfulBreeze commented:
Also only AIs have KDA, so this KDA tracking code should probably go to ObjAIBase

danil179 committed:
Changed OnKillMinion to OnKill and added type safety

danil179 commented:
Changes completed.

PR ready for merge.

HarmfulBreeze commented on GameServerLib/Logic/GameObjects/AttackableUnits/AttackableUnit.cs:
```diff
@@ -114,7 +114,7 @@ public virtual void Die(AttackableUnit killer)
                 }
 
-                cKiller.OnKillMinion(this as Minion);
+                cKiller.OnKill(this);
```
Maybe just change it to
```cs
if ((killer != null) && (killer is Champion))
{
    ((Champion)killer).OnKill(this)
}
```
Would simplify the code a lot.
The rest looks 👌 to me

danil179 replied:
It will not return if killer is not a champion, and that will change the flow of code to change the IsDashing variable from true to false, and I'm not sure when it should be changed.

HarmfulBreeze replied:
I'd write this:
```cs
if ((killer != null) && (killer is Champion))
{
    ((Champion)killer).OnKill(this)
}
IsDashing = false;
```
Because you want to cancel dashes even when there is no ``Champion`` who killed him (using the death command for example, or being killed solely by minions).
(also this check on IsDashing is redundant so it gets removed too)

danil179 replied:
Ok, will change.

danil179 committed:
Code cleaned

danil179 committed:
Fixed 2 bugs
gold was not added on kill of other champion
if player got killed from turret it was counted that he killed himself

danil179 commented:
Fixed some bugs.

PR ready for merge.

HarmfulBreeze commented on GameServerLib/Logic/GameObjects/AttackableUnits/AI/Champion.cs:
```diff
@@ -535,7 +535,7 @@ public override void TakeDamage(AttackableUnit attacker, float damage, DamageTyp
 
             _championHitFlagTimer = 15 * 1000; //15 seconds timer, so when you get executed the last enemy champion who hit you gets the gold
-            _playerHitId = NetId;
+            _playerHitId = attacker.NetId;
```
That one is a funny one
